### PR TITLE
[Github Actions] - Cut Release Branch Workflow

### PR DESCRIPTION
**Pull request by benWize**
As part of the [Beam Release Process](https://beam.apache.org/contribute/release-guide/), the following PR includes the workflow to run the [cut_release_branch.sh](https://github.com/apache/beam/blob/master/release/src/main/scripts/cut_release_branch.sh) script described in [step 5. Use cut_release_branch.sh to cut a release branch
](https://beam.apache.org/contribute/release-guide/#use-cut_release_branchsh-to-cut-a-release-branch)

Files:
- Cut Release Branch: .github/workflows/cut_release_branch.yml
- Updated CI.md file

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: .github/workflows/cut_release_branch.yml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+      MASTER_BRANCH: master
+      NEXT_RELEASE: ${{ github.event.inputs.NEXT_VERSION }}
+      SCRIPT_DIR: ./release/src/main/scripts
+    steps:
+      - name: Validate Next Version
+        run: |
+          if [[ $NEXT_RELEASE =~ ([0-9]\.[0-9]*\.[0-9]) ]]; then
+            NEXT_VERSION_IN_BASE_BRANCH=${BASH_REMATCH[1]}
+          fi
+          if [[ -z "$NEXT_VERSION_IN_BASE_BRANCH" ]]; then
```
Comment (damccorm) [inline .github/workflows/cut_release_branch.yml:47]: Any reason not to just use an `else`?

Comment (damccorm) [reply .github/workflows/cut_release_branch.yml:47]: Then you could just echo that directly into the GITHUB_ENV as well.

File: .github/workflows/cut_release_branch.yml [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            echo "The input for NEXT_RELEASE does not match a valid format [0-9].[0-9].[0-9]"
+            exit 1
+          fi
+          echo "NEXT_VERSION_IN_BASE_BRANCH=${NEXT_VERSION_IN_BASE_BRANCH}" >> $GITHUB_ENV
+      - name: Echo value
+        run: |
+          echo "NEXT_VERSION_IN_BASE_BRANCH: $NEXT_VERSION_IN_BASE_BRANCH"
```
Comment (damccorm) [inline .github/workflows/cut_release_branch.yml:54]: Do we need this? Why not just print it in the previous step if you want that information?

File: .github/workflows/cut_release_branch.yml [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+          bash "${SCRIPT_DIR}/set_version.sh" "${NEXT_VERSION_IN_BASE_BRANCH}"
+          echo "==============Update master branch as following================"
+          git diff
+          echo "==============================================================="
+      - name: Commit and Push to master branch files with Next Version
+        run: |
+          git add buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy
+          git add gradle.properties
+          git add sdks/python/apache_beam/version.py
+          git add sdks/go/pkg/beam/core/core.go
```
Comment (damccorm) [inline .github/workflows/cut_release_branch.yml:72]: Can this just be `git add *` or does `set_version.sh` change more than just that?

File: .github/workflows/cut_release_branch.yml [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    runs-on: [self-hosted, ubuntu-20.04]
+    env:
+      RELEASE: ${{ github.event.inputs.RELEASE_VERSION }}
+    steps:
+      - name: Validate Release Version
+        run: |
+          if [[ ${RELEASE} =~ ([0-9]\.[0-9]*\.[0-9]) ]]; then
+            RELEASE_VERSION=${BASH_REMATCH[1]}
+          fi
+          if [[ -z "$RELEASE_VERSION" ]]; then
```
Comment (damccorm) [inline .github/workflows/cut_release_branch.yml:86]: Same comments about `else`

File: .github/workflows/cut_release_branch.yml [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      - name: Commit and Push to master branch files with Next Version
+        run: |
+          git add buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy
+          git add gradle.properties
+          git add sdks/python/apache_beam/version.py
+          git add sdks/go/pkg/beam/core/core.go
+          git commit -m "Moving to ${NEXT_VERSION_IN_BASE_BRANCH}-SNAPSHOT on master branch."
+          git push origin ${MASTER_BRANCH}
+
+  update_release_branch:
```
Comment (damccorm) [inline .github/workflows/cut_release_branch.yml:76]: Does the ordering matter here? Should `update_master` have a dependency on this step and run after it?

**Comment by damccorm** [comment]
I reviewed this for correctness/style, but I'm not super-familiar with the actual steps needed to run it. @kennknowles could you take a look as the next release manager?

File: .github/workflows/cut_release_branch.yml (hunk 1 above)
Comment (benWize) [reply .github/workflows/cut_release_branch.yml]: Thanks for your suggestion, I changed it to use else

File: .github/workflows/cut_release_branch.yml (hunk 2 above)
Comment (benWize) [reply .github/workflows/cut_release_branch.yml]: Removed it

File: .github/workflows/cut_release_branch.yml (hunk 3 above)
Comment (benWize) [reply .github/workflows/cut_release_branch.yml]: I think we can use git add *, changed it

File: .github/workflows/cut_release_branch.yml (hunk 5 above)
Comment (benWize) [reply .github/workflows/cut_release_branch.yml:69]: Added as a dependency

File: .github/workflows/cut_release_branch.yml [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    needs: update_release_branch
+    runs-on: [self-hosted, ubuntu-20.04]
+    env:
+      MASTER_BRANCH: master
+      NEXT_RELEASE: ${{ github.event.inputs.NEXT_VERSION }}
+      SCRIPT_DIR: ./release/src/main/scripts
+    steps:
+      - name: Validate Next Version
+        run: |
+          if [[ $NEXT_RELEASE =~ ([0-9]\.[0-9]*\.[0-9]) ]]; then
```
Comment (kennknowles) [inline .github/workflows/cut_release_branch.yml:45]: The regex should be `[0-9]+\.[0-9]+\.[0-9]+` here and in the error message below.

File: .github/workflows/cut_release_branch.yml (hunk 5 above)
Comment (kennknowles) [reply .github/workflows/cut_release_branch.yml:69]: Ordering does matter.`update_master` should run first (the opposite of what you have now). If you cut the branch first, then a PR could merge after the cut point and you have commits with the wrong SNAPSHOT version on them.

The branch should be cut on the parent commit of the one from `update_master`. In other words, the last commit on master with the SNAPSHOT version.

In fact, the two versions are really symmetrical. You can either think of it as a release branch being cut and master updating the version _or_ you can think of it as the release branch being a continuation of the main branch, and a new `master` branch being created with the new snapshot version. The latter interpretation is actually very helpful for me to think about things.

So in fact updating the version on the master branch should be committed first to step any new PRs being merged. If any PRs were merged with the old SNAPSHOT version, the release branch should cut from their merge commit.

**Comment by kennknowles** [comment]
The invariants we want to hold for the commit history:

 - Any commit on any branch should have a -SNAPSHOT version
 - RCs and releases should be on tags that are not on branches (because they are terminal)
 - For any release tag X.Y.Z every commit included in the tag has a <= version.
 - For any commit with X.Y.Z-SNAPSHOT version, it is included in the RC and release tags for X.Y.Z

**Comment by benWize** [comment]
Thanks, @kennknowles, I changed the regex and the job dependency

**Comment by kennknowles** [comment]
Thinking about how this runs, it will already have chosen a commit so the two jobs do not have any dependency on each other. They will have the same effect regardless.

I believe the best way to do it would be to advance the master branch version in a separate step and then explicitly pass the best commit to the cut release branch.

But for now we should just merge and go ahead and I will try this action out for cutting the current branch. It is as good as what we do today (unless a release manager is careful and does it manually)

**Review by kennknowles — APPROVED** [review]

**Comment by kennknowles** [comment]
I see. I will cut the release branch manually.

**Comment by kennknowles** [comment]
Though this one does not require a self-hosted runner.

**Pull request merged**